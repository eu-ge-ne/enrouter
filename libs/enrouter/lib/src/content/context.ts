import { createContext, useContext } from "react";

import type { Content } from "./mod.js";

const ContentContext = createContext<Content>({
  match: {
    route: {
      path: "",
      test: {
        keys: [],
        pattern: new RegExp(""),
      },
      modules: [],
      loaded: false,
      elements: {},
    },
    location: "",
    isFull: false,
    params: {},
  },
});

export const ContentProvider = ContentContext.Provider;

export function useContent(): Content {
  return useContext(ContentContext);
}
