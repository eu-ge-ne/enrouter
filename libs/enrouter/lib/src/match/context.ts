import { createContext, useContext } from "react";

import type { Match } from "./mod.js";

const MatchContext = createContext<Match>({
  route: {
    path: "",
    test: { keys: [], pattern: new RegExp("") },
    modules: [],
    loaded: false,
    elements: {},
  },

  isFull: false,
  location: "",
  params: {},
});

export const MatchProvider = MatchContext.Provider;

export function useMatch(): Match {
  return useContext(MatchContext);
}
