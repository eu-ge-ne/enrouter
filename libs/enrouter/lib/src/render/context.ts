import { createContext } from "react";

import type { RouteNodes } from "./mod.js";

export const RouteRenderContext = createContext<RouteNodes>({
  match: {
    handler: {
      route: {
        path: "",
        mod: [],
        link: [[], []],
      },
      test: { keys: [], pattern: new RegExp("") },
      modules: [],
    },
    location: "",
    isFull: false,
    params: {},
  },
});
