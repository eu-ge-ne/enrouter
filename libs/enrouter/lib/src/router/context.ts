import { createContext, useContext } from "react";

import type { Route } from "#lib/route/mod.js";

export interface TRouterContext {
  routes: Route;
  location: string;
  navigate: (to: string) => void;
  ctx: unknown;
}

const RouterContext = createContext<TRouterContext>({
  routes: {
    path: "",
    test: { keys: [], pattern: new RegExp("") },
    modules: [],
    loaded: false,
    elements: {},
  },
  location: "",
  navigate: () => undefined,
  ctx: undefined,
});

export const RouterProvider = RouterContext.Provider;

export function useRouter(): TRouterContext {
  return useContext(RouterContext);
}
