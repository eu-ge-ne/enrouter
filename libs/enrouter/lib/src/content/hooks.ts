import type { Match } from "#lib/match/mod.js";
import { useContent } from "./context.js";

export function useMatch(): Match {
  return useContent().match;
}

export function usePath(path: string): Match | undefined {
  let x = useContent().last;
  while (x) {
    if (x.match.route.path === path) {
      return x.match;
    }
    x = x.prev;
  }
}

export interface UseActiveParams<T> {
  path: string;
  loose?: boolean;
  value: [T, T];
}

export function useActive<T>({ path, loose, value }: UseActiveParams<T>): T {
  const match = usePath(path);

  const isActive = loose ? match !== undefined : Boolean(match?.isFull);

  return isActive ? value[0] : value[1];
}
