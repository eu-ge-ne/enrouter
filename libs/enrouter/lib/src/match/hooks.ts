import type { Match } from "./mod.js";
import { useMatch } from "./context.js";

export function useRoot(): Match | undefined {
  let match = useMatch()?.last;

  do {
    if (match?.route.elements._root) {
      return match;
    }
    match = match?.prev;
  } while (match);
}

export function usePath(path: string): Match | undefined {
  let match = useMatch()?.first;

  while (match) {
    if (match.route?.path === path) {
      return match;
    }

    match = match.next;
  }
}

export interface UseActiveParams<T> {
  path: string;
  loose?: boolean;
  value: [T, T];
}

export function useActive<T>({ path, loose, value }: UseActiveParams<T>): T {
  const match = usePath(path);

  const isActive = loose ? match !== undefined : Boolean(match?.full);

  return isActive ? value[0] : value[1];
}
