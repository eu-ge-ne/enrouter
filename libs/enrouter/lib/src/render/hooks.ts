import { useContext } from "react";

import { RouteRenderContext } from "./context.js";
import type { RouteMatch } from "#lib/match/mod.js";

export function useMatch(): RouteMatch {
  return useContext(RouteRenderContext).match;
}

export function usePath(path: string): RouteMatch | undefined {
  let nodes = useContext(RouteRenderContext).last;
  while (nodes) {
    if (nodes.match.route.path === path) {
      return nodes.match;
    }
    nodes = nodes.prev;
  }
}

export interface UseActiveParams<T> {
  path: string;
  loose?: boolean;
  value: [T, T];
}

export function useActive<T>({ path, loose, value }: UseActiveParams<T>): T {
  const match = usePath(path);

  const isActive = loose ? match !== undefined : Boolean(match?.isFull);

  return isActive ? value[0] : value[1];
}
