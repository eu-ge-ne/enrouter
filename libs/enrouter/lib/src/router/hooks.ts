import { useRouterDynamic } from "./context.js";

export function useLocation(): string {
  const { location } = useRouterDynamic();
  return location;
}
