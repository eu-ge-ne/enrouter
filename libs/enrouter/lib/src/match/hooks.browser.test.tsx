import type { FC, PropsWithChildren } from "react";
import { describe, test, expect } from "vitest";
import { render } from "vitest-browser-react";
import * as regexparam from "regexparam";

import type { Match } from "./mod.js";
import { MatchProvider } from "./context.js";
import { usePath, useActive } from "./hooks.js";

const wrapperId = "test-wrapper";

const wrapper: FC<PropsWithChildren> = ({ children }) => (
  <div data-testid={wrapperId}>{children}</div>
);

describe("match", () => {
  describe("usePath", () => {
    test("with 2 matches", async () => {
      const context: Match = {
        route: {
          path: "/",
          test: regexparam.parse("/", true),
          modules: [],
          loaded: true,
          elements: {
            this: {
              Main: <div>Layout /</div>,
            },
          },
        },
        isFull: false,
        location: "/",
        params: {},

        next: {
          route: {
            path: "/abc",
            test: regexparam.parse("/abc", true),
            modules: [],
            loaded: true,
            elements: {
              this: {
                Main: <div>Layout /abc</div>,
              },
            },
          },
          isFull: true,
          location: "/abc",
          params: {},
        },
      };

      context.first = context;

      const Test: FC = () => {
        const match = usePath("/abc");
        return Object.values(match?.route.elements?.this ?? {});
      };

      const screen = render(
        <MatchProvider value={context}>
          <Test />
        </MatchProvider>,
        { wrapper },
      );

      await expect.element(screen.getByTestId(wrapperId)).toBeVisible();

      expect(screen.container).toMatchSnapshot();
    });
  });

  describe("useActive", () => {
    test("with 2 matches", async () => {
      const context: Match = {
        route: {
          path: "/",
          test: regexparam.parse("/", true),
          modules: [],
          loaded: true,
          elements: {
            this: {
              Main: <div>Layout /</div>,
            },
          },
        },
        isFull: false,
        location: "/",
        params: {},

        next: {
          route: {
            path: "/abc",
            test: regexparam.parse("/abc", true),
            modules: [],
            loaded: true,
            elements: {
              this: {
                Main: <div>Layout /abc</div>,
              },
            },
          },
          isFull: true,
          location: "/abc",
          params: {},
        },
      };

      context.first = context;

      const Test: FC = () => {
        const match1 = useActive({
          path: "/",
          value: ["/ is active", "/ is not active"],
        });
        const match2 = useActive({
          path: "/abc",
          value: ["/abc is active", "/abc is not active"],
        });
        const match3 = useActive({
          path: "/",
          loose: true,
          value: ["/ is active (loose)", "/ is not active (loose)"],
        });
        const match4 = useActive({
          path: "/xyz",
          value: ["/xyz is active", "/xyz is not active"],
        });

        return (
          <div>
            <p>{match1}</p>
            <p>{match2}</p>
            <p>{match3}</p>
            <p>{match4}</p>
          </div>
        );
      };

      const screen = render(
        <MatchProvider value={context}>
          <Test />
        </MatchProvider>,
        { wrapper },
      );

      await expect.element(screen.getByTestId(wrapperId)).toBeVisible();

      expect(screen.container).toMatchSnapshot();
    });
  });
});
