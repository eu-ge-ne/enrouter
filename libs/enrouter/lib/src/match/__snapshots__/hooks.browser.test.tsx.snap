// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`match > useActive > with 2 matches 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    <div>
      <p>
        / is not active
      </p>
      <p>
        /abc is active
      </p>
      <p>
        / is active (loose)
      </p>
      <p>
        /xyz is not active
      </p>
    </div>
  </div>
</div>
`;

exports[`match > usePath > with 2 matches 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    <div>
      Layout /abc
    </div>
  </div>
</div>
`;
